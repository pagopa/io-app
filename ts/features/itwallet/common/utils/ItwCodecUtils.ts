/**
 * Copied from https://github.com/gcanti/io-ts-types/blob/master/src/JsonFromString.ts
 */

/**
 * @since 0.5.14
 */
import * as t from "io-ts";

/**
 * Copied from `fp-ts/Either` module.
 *
 * @since 0.5.14
 */
export type Json = boolean | number | string | null | JsonArray | JsonRecord;

/**
 * @since 0.5.14
 */
export interface JsonRecord {
  readonly [key: string]: Json;
}

/**
 * @since 0.5.14
 */
export type JsonArray = ReadonlyArray<Json>;

/**
 * @since 0.5.15
 */
export const JsonArray: t.Type<JsonArray> = t.recursion("JsonArray", () =>
  t.readonlyArray(Json)
);

/**
 * @since 0.5.15
 */
export const JsonRecord: t.Type<JsonRecord> = t.recursion("JsonRecord", () =>
  t.record(t.string, Json)
);

/**
 * @since 0.5.15
 */
export const Json: t.Type<Json> = t.union(
  [t.boolean, t.number, t.string, t.null, JsonArray, JsonRecord],
  "Json"
);

/**
 * @since 0.5.14
 */
export const JsonFromString = new t.Type<Json, string, string>(
  "JsonFromString",
  Json.is,
  (s, c) => {
    try {
      return t.success(JSON.parse(s));
    } catch (e) {
      return t.failure(s, c);
    }
  },
  json => JSON.stringify(json)
);
