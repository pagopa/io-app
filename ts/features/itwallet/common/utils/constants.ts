/**
 * Defines the number of asterisks used to mask the value of claims in the credential details
 */
export const HIDDEN_CLAIM_TEXT = "******";
