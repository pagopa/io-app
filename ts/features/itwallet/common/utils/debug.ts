/**
 * Debug utilities
 */

import { ParsedCredential } from "./itwTypesUtils";

/**
 * Removes very long claims (like Base64 images) from the parsed credential. Usefull to print for debug purposes
 */
export const getHumanReadableParsedCredential = (
  parsedCredential: ParsedCredential
): ParsedCredential =>
  Object.fromEntries(
    Object.entries(parsedCredential).map(([key, { name, value }]) => [
      key,
      {
        name,
        value:
          typeof value === "string" && value.length > 100
            ? `${value.slice(0, 100)}...`
            : value
      }
    ])
  );
