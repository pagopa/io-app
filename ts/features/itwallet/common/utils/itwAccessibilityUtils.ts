import I18n from "../../../../i18n";
import { ItwCredentialStatus } from "./itwTypesUtils";

export const accessibilityLabelByStatus: {
  [key in ItwCredentialStatus]?: string;
} = {
  invalid: I18n.t("features.itWallet.card.status.invalid"),
  expired: I18n.t("features.itWallet.card.status.expired"),
  jwtExpired: I18n.t("features.itWallet.card.status.verificationExpired"),
  expiring: I18n.t("features.itWallet.card.status.expiring"),
  jwtExpiring: I18n.t("features.itWallet.card.status.verificationExpiring")
};
