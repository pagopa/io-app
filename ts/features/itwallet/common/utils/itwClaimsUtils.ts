/**
 * Utility functions for working with credential claims.
 */

import * as t from "io-ts";
import { PatternString } from "@pagopa/ts-commons/lib/strings";
import { patternDateFromString } from "@pagopa/ts-commons/lib/dates";
import { pipe } from "fp-ts/lib/function";
import * as O from "fp-ts/lib/Option";
import * as E from "fp-ts/lib/Either";
import { Locales } from "../../../../../locales/locales";
import I18n from "../../../../i18n";
import { ParsedCredential, StoredCredential } from "./itwTypesUtils";
import { CredentialCatalogDisplay } from "./itwMocksUtils";

/**
 *
 *
 *
 * CLAIMS MANIPULATION UTILS
 *
 *
 *
 */

/**
 * Retrieves the organization name from the evidence claim of the given credential.
 * If the evidence claim is not present or cannot be decoded, a fallback value is returned.
 * @param credential - The parsed credential object.
 * @returns The organization name from the evidence claim or a fallback value.
 */
export const getEvidenceOrganizationName = (credential: ParsedCredential) =>
  pipe(
    credential.evidence,
    O.fromNullable,
    O.fold(
      () => I18n.t("features.itWallet.generic.placeholders.organizationName"),
      evidence =>
        pipe(
          evidence.value,
          EvidenceClaim.decode,
          E.fold(
            () =>
              I18n.t("features.itWallet.generic.placeholders.organizationName"),
            some => some[0].record.source.organization_name
          )
        )
    )
  );

/**
 * Type for each claim to be displayed.
 */
export type ClaimDisplayFormat = {
  id: string;
  label: string;
  value: unknown;
};

/**
 * Parses the claims from the credential.
 * For each Record entry it maps the key and the attribute value to a label and a value.
 * The label is taken from the attribute name which is either a string or a record of locale and string.
 * If the type of the attribute name is string then when take it's value because locales have not been set.
 * If the type of the attribute name is record then we take the value of the locale that matches the current locale.
 * If there's no locale that matches the current locale then we take the attribute key as the name.
 * The value is taken from the attribute value.
 * @param parsedCredential - the parsed credential.
 * @returns the array of {@link ClaimDisplayFormat} of the credential contained in its configuration schema.
 */
export const parseClaims = (
  parsedCredential: ParsedCredential
): Array<ClaimDisplayFormat> =>
  Object.entries(parsedCredential).map(([key, attribute]) => {
    const attributeName =
      typeof attribute.name === "string"
        ? attribute.name
        : attribute.name?.[getClaimsFullLocale()] || key;

    return { label: attributeName, value: attribute.value, id: key };
  });

/**
 * Sorts the parsedCredential according to the order of the displayData.
 * If the order is not available, the schema is returned as is.
 * @param parsedCredential - the parsed credential.
 * @param order - the order of the displayData.
 * @returns a new parsedCredential sorted according to the order of the displayData.
 */
export const sortClaims = (
  order: CredentialCatalogDisplay["order"],
  parsedCredential: ParsedCredential
) =>
  order
    ? Object.fromEntries(
        Object.entries(parsedCredential)
          .slice()
          .sort(([key1], [key2]) => order.indexOf(key1) - order.indexOf(key2))
      )
    : parsedCredential;

/**
 *
 *
 *
 * CLAIMS LOCALE UTILS
 *
 *
 *
 */

/**
 * Enum for the claims locales.
 * This is used to get the correct locale for the claims.
 * Currently the only supported locales are it-IT and en-US.
 */
enum ClaimsLocales {
  it = "it-IT",
  en = "en-US"
}

/**
 * Map from the app locales to the claims locales.
 * Currently en is mapped to en-US and it to it-IT.
 */
const localeToClaimsLocales = new Map<Locales, ClaimsLocales>([
  ["it", ClaimsLocales.it],
  ["en", ClaimsLocales.en]
]);

/**
 * Helper function to get a full claims locale locale from the current app locale.
 * @returns a enum value for the claims locale.
 */
export const getClaimsFullLocale = (): ClaimsLocales =>
  localeToClaimsLocales.get(I18n.currentLocale()) ?? ClaimsLocales.it;

/**
 *
 *
 *
 * IO-TS DECODER FOR THE CLAIMS
 *
 *
 *
 */

/**
 * Regex for the date format which is used to validate the date claim as ISO 8601:2004 YYYY-MM-DD format.
 */
const DATE_FORMAT_REGEX = "^\\d{4}-\\d{2}-\\d{2}$";

/**
 * Regex for the picture URL format which is used to validate the image claim as a base64 encoded png image.
 */
const PICTURE_URL_REGEX = "^data:image\\/png;base64,";

/**
 * io-ts decoder for the date claim field of the credential.
 * The date format is checked against the regex dateFormatRegex, which is currenlty mocked.
 * This is needed because a generic date decoder would accept invalid dates like numbers,
 * thus decoding properly and returning a wrong claim item to be displayed.
 */
export const DateClaim = patternDateFromString(DATE_FORMAT_REGEX, "DateClaim");

/**
 * io-ts decoder for the evidence claim field of the credential.
 */
export const EvidenceClaim = t.array(
  t.type({
    type: t.string,
    record: t.type({
      type: t.string,
      source: t.type({
        organization_name: t.string,
        organization_id: t.string,
        country_code: t.string
      })
    })
  })
);

/**
 * io-ts decoder for the place of birth claim field of the credential.
 */
export const PlaceOfBirthClaim = t.type({
  country: t.string,
  locality: t.string
});
export type PlaceOfBirthClaimType = t.TypeOf<typeof PlaceOfBirthClaim>;

/**
 * io-ts decoder for the mDL driving privileges
 */
export const DrivingPrivilegesClaim = t.type({
  issue_date: t.string,
  vehicle_category_code: t.string,
  expiry_date: t.string
});
export type DrivingPrivilegesClaimType = t.TypeOf<
  typeof DrivingPrivilegesClaim
>;

/**
 * Alias for the string fallback of the claim field of the credential.
 */
export const PlainTextClaim = t.string;

export const ImageClaim = PatternString(PICTURE_URL_REGEX);

/**
 * Decoder type for the claim field of the credential.
 * It includes all the possible types of claims and fallbacks to string.
 * To add more custom objects to the union:
 * t.string.pipe(JsonFromString).pipe(t.union([PlaceOfBirthClaim, PlaceOfBirthClaim]))
 */
export const ClaimValue = t.union([
  // Parse an object representing the place of birth
  PlaceOfBirthClaim,
  // Parse an object representing a mDL driving privileges
  DrivingPrivilegesClaim,
  // Parse an object representing the claim evidence
  EvidenceClaim,
  // Otherwise parse a date
  DateClaim,
  // Otherwise parse an image
  ImageClaim,
  // Otherwise fallback to string
  PlainTextClaim
]);

type ClaimSection =
  | "personalData"
  | "documentData"
  | "licenseData"
  | "noSection";

export type DateClaimConfig = Partial<{
  iconVisible: boolean;
  expirationBadgeVisible: boolean;
}>;

/**
 * Hardcoded claims sections: currently it's not possible to determine how to group claims from the credential.
 * The order of the claims doesn't matter here, the credential's `displayData` order wins.
 * Claims that are present here but not in the credential are safely ignored.
 */
const sectionsByClaim: Record<string, ClaimSection> = {
  // Personal data claims
  given_name: "personalData",
  family_name: "personalData",
  birthdate: "personalData",
  place_of_birth: "personalData",
  tax_id_code: "personalData",
  tax_id_number: "personalData",
  portrait: "personalData",
  sex: "personalData",

  // Document data claims
  issue_date: "documentData",
  expiry_date: "documentData",
  expiration_date: "documentData",
  document_number: "documentData",

  // Driving license claims
  driving_privileges: "licenseData"
};

export const dateClaimsConfig: Record<string, DateClaimConfig> = {
  issue_date: { iconVisible: true },
  expiry_date: { iconVisible: true, expirationBadgeVisible: true },
  expiration_date: { iconVisible: true, expirationBadgeVisible: true }
};

export const previewDateClaimsConfig: DateClaimConfig = {
  iconVisible: false,
  expirationBadgeVisible: false
};

/**
 * Groups claims in a credential according to {@link sectionsByClaim}.
 * Claims are assigned to the designated section in the order specified by the credential's `displayData`.
 * Claims without a section are assigned to the key `noSection` so they can be rendered separately.
 * @param credential
 * @returns
 */
export const groupCredentialClaims = (credential: StoredCredential) => {
  const claims = parseClaims(credential.parsedCredential);

  return claims.reduce((acc, claim) => {
    const section = sectionsByClaim[claim.id] || "noSection";
    return {
      ...acc,
      [section]: (acc[section] || []).concat(claim)
    };
  }, {} as Record<ClaimSection, ReadonlyArray<ClaimDisplayFormat>>);
};
