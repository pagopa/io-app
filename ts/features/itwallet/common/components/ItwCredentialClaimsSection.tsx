import React, { useState } from "react";
import { StyleSheet, View } from "react-native";
import { IconButton, H6, Divider } from "@pagopa/io-app-design-system";
import I18n from "../../../../i18n";
import { ClaimDisplayFormat } from "../utils/itwClaimsUtils";
import { ItwCredentialClaim } from "./ItwCredentialClaim";

type Props = {
  title: string;
  claims: ReadonlyArray<ClaimDisplayFormat>;
  canHideValues?: boolean;
};

export const ItwCredentialClaimsSection = ({
  title,
  canHideValues,
  claims
}: Props) => {
  const [valuesHidden, setValuesHidden] = useState(false);

  const renderHideValuesToggle = () => (
    <IconButton
      testID="toggle-claim-visibility"
      icon={valuesHidden ? "eyeHide" : "eyeShow"}
      onPress={() => setValuesHidden(x => !x)}
      accessibilityLabel={I18n.t(
        valuesHidden
          ? "features.itWallet.presentation.credentialDetails.actions.showClaimValues"
          : "features.itWallet.presentation.credentialDetails.actions.hideClaimValues"
      )}
    />
  );

  return (
    <View>
      <View style={styles.header}>
        <H6 color="grey-700">{title}</H6>
        {canHideValues && renderHideValuesToggle()}
      </View>
      <View>
        {claims.map((claim, index) => (
          <React.Fragment key={claim.id}>
            {index !== 0 && <Divider />}
            <ItwCredentialClaim claim={claim} hidden={valuesHidden} />
          </React.Fragment>
        ))}
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  header: {
    justifyContent: "space-between",
    flexDirection: "row"
  }
});
