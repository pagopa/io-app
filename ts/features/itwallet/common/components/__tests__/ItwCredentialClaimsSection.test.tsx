import { render, fireEvent } from "@testing-library/react-native";
import * as React from "react";
import { ItwCredentialClaimsSection } from "../ItwCredentialClaimsSection";
import { ClaimDisplayFormat } from "../../utils/itwClaimsUtils";

describe("ItwCredentialClaimsSection", () => {
  const claims: Array<ClaimDisplayFormat> = [
    { id: "name", label: "Nome", value: "Luigi" },
    { id: "surname", label: "Cognome", value: "Rossi" },
    { id: "age", label: "EtÃ ", value: 20 }
  ];

  it("should match the snapshot when claims are visible", () => {
    const component = render(
      <ItwCredentialClaimsSection
        title="Personal data"
        claims={claims}
        canHideValues
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("should match the snapshot when claims are hidden", () => {
    const component = render(
      <ItwCredentialClaimsSection
        title="Personal data"
        claims={claims}
        canHideValues
      />
    );
    const toggleButton = component.queryByTestId("toggle-claim-visibility");

    if (!toggleButton) {
      fail("Toggle button not found");
    }

    fireEvent(toggleButton, "onPress");
    expect(component).toMatchSnapshot();
  });
});
