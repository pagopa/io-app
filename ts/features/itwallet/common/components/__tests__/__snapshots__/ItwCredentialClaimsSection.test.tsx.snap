// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItwCredentialClaimsSection should match the snapshot when claims are hidden 1`] = `
<View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      allowFontScaling={false}
      color="grey-700"
      defaultColor="black"
      defaultWeight="Semibold"
      font="TitilliumSansPro"
      fontStyle={
        {
          "fontSize": 18,
          "lineHeight": 25,
        }
      }
      style={
        [
          {
            "fontSize": 18,
            "lineHeight": 25,
          },
          {
            "color": "#555C70",
            "fontFamily": "Titillium Sans Pro",
            "fontStyle": "normal",
            "fontWeight": "600",
          },
        ]
      }
      weight="Semibold"
    >
      Personal data
    </Text>
    <View
      accessibilityLabel="Show claim values"
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={8}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="toggle-claim-visibility"
    >
      <View
        style={
          [
            {
              "height": 24,
              "width": 24,
            },
            {
              "alignItems": "center",
            },
            {
              "justifyContent": "center",
            },
            {
              "transform": [
                {
                  "scale": undefined,
                },
              ],
            },
          ]
        }
      >
        <RNSVGSvgView
          accessibilityElementsHidden={true}
          accessibilityLabel=""
          accessible={false}
          align="xMidYMid"
          animatedProps={
            {
              "color": undefined,
            }
          }
          bbHeight={24}
          bbWidth={24}
          focusable={false}
          height={24}
          importantForAccessibility="no-hide-descendants"
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "color": "#0073E6",
              },
              {
                "flex": 0,
                "height": 24,
                "width": 24,
              },
            ]
          }
          tintColor="#0073E6"
          vbHeight={24}
          vbWidth={24}
          width={24}
        >
          <RNSVGGroup
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
          >
            <RNSVGPath
              clipRule={0}
              d="M1.707.293A1 1 0 0 0 .293 1.707l4.483 4.484A13.063 13.063 0 0 0 .63 10.696a2.686 2.686 0 0 0 0 2.609A12.997 12.997 0 0 0 12 20c1.931 0 3.763-.42 5.41-1.176l4.883 4.883a1 1 0 0 0 1.414-1.414l-22-22Zm19.099 16.27c1.02-.939 1.888-2.039 2.566-3.259.45-.812.45-1.797 0-2.609A12.997 12.997 0 0 0 12.001 4c-1.15 0-2.266.15-3.329.43l1.691 1.691a10.997 10.997 0 0 1 11.26 5.545.686.686 0 0 1 0 .667 11.05 11.05 0 0 1-2.233 2.815l1.416 1.416Zm-3.856-3.855a5 5 0 0 0-5.658-5.658l2.804 2.803.05.051 2.804 2.804Zm-14.573-1.04a11.055 11.055 0 0 1 3.846-4.03l1.605 1.605a5 5 0 0 0 6.929 6.929l1.125 1.124A10.977 10.977 0 0 1 12 18a10.997 10.997 0 0 1-9.623-5.666.686.686 0 0 1 0-.667ZM9 12c0-.463.105-.902.292-1.293l4.001 4A3 3 0 0 1 9 12Z"
              fill={
                {
                  "type": 2,
                }
              }
              fillRule={0}
              propList={
                [
                  "fill",
                  "fillRule",
                ]
              }
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
    </View>
  </View>
  <View>
    <View
      accessibilityLabel="Nome Hidden"
      accessible={true}
      style={
        {
          "marginHorizontal": -24,
          "paddingHorizontal": 24,
          "paddingVertical": 12,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <View
            accessibilityElementsHidden={false}
            accessible={true}
            importantForAccessibility="yes"
          >
            <Text
              allowFontScaling={false}
              color="grey-700"
              defaultColor="blue"
              defaultWeight="Bold"
              font="TitilliumSansPro"
              fontStyle={
                {
                  "fontSize": 14,
                  "lineHeight": 21,
                }
              }
              style={
                [
                  {
                    "fontSize": 14,
                    "lineHeight": 21,
                  },
                  {
                    "color": "#555C70",
                    "fontFamily": "Titillium Sans Pro",
                    "fontStyle": "normal",
                    "fontWeight": "400",
                  },
                ]
              }
              weight="Regular"
            >
              Nome
            </Text>
            <Text
              allowFontScaling={false}
              color="black"
              defaultColor="black"
              defaultWeight="Semibold"
              font="TitilliumSansPro"
              fontStyle={
                {
                  "fontSize": 18,
                  "lineHeight": 25,
                }
              }
              numberOfLines={2}
              style={
                [
                  {
                    "fontSize": 18,
                    "lineHeight": 25,
                  },
                  {
                    "color": "#0E0F13",
                    "fontFamily": "Titillium Sans Pro",
                    "fontStyle": "normal",
                    "fontWeight": "600",
                  },
                ]
              }
              weight="Semibold"
            >
              ******
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "#D2D6E3",
          "height": 0.5,
        }
      }
    />
    <View
      accessibilityLabel="Cognome Hidden"
      accessible={true}
      style={
        {
          "marginHorizontal": -24,
          "paddingHorizontal": 24,
          "paddingVertical": 12,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <View
            accessibilityElementsHidden={false}
            accessible={true}
            importantForAccessibility="yes"
          >
            <Text
              allowFontScaling={false}
              color="grey-700"
              defaultColor="blue"
              defaultWeight="Bold"
              font="TitilliumSansPro"
              fontStyle={
                {
                  "fontSize": 14,
                  "lineHeight": 21,
                }
              }
              style={
                [
                  {
                    "fontSize": 14,
                    "lineHeight": 21,
                  },
                  {
                    "color": "#555C70",
                    "fontFamily": "Titillium Sans Pro",
                    "fontStyle": "normal",
                    "fontWeight": "400",
                  },
                ]
              }
              weight="Regular"
            >
              Cognome
            </Text>
            <Text
              allowFontScaling={false}
              color="black"
              defaultColor="black"
              defaultWeight="Semibold"
              font="TitilliumSansPro"
              fontStyle={
                {
                  "fontSize": 18,
                  "lineHeight": 25,
                }
              }
              numberOfLines={2}
              style={
                [
                  {
                    "fontSize": 18,
                    "lineHeight": 25,
                  },
                  {
                    "color": "#0E0F13",
                    "fontFamily": "Titillium Sans Pro",
                    "fontStyle": "normal",
                    "fontWeight": "600",
                  },
                ]
              }
              weight="Semibold"
            >
              ******
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "#D2D6E3",
          "height": 0.5,
        }
      }
    />
    <View
      accessibilityLabel="EtÃ  Attributo non riconosciuto"
      accessible={true}
      style={
        {
          "marginHorizontal": -24,
          "paddingHorizontal": 24,
          "paddingVertical": 12,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <View
            accessibilityElementsHidden={false}
            accessible={true}
            importantForAccessibility="yes"
          >
            <Text
              allowFontScaling={false}
              color="grey-700"
              defaultColor="blue"
              defaultWeight="Bold"
              font="TitilliumSansPro"
              fontStyle={
                {
                  "fontSize": 14,
                  "lineHeight": 21,
                }
              }
              style={
                [
                  {
                    "fontSize": 14,
                    "lineHeight": 21,
                  },
                  {
                    "color": "#555C70",
                    "fontFamily": "Titillium Sans Pro",
                    "fontStyle": "normal",
                    "fontWeight": "400",
                  },
                ]
              }
              weight="Regular"
            >
              EtÃ 
            </Text>
            <Text
              allowFontScaling={false}
              color="black"
              defaultColor="black"
              defaultWeight="Semibold"
              font="TitilliumSansPro"
              fontStyle={
                {
                  "fontSize": 18,
                  "lineHeight": 25,
                }
              }
              numberOfLines={2}
              style={
                [
                  {
                    "fontSize": 18,
                    "lineHeight": 25,
                  },
                  {
                    "color": "#0E0F13",
                    "fontFamily": "Titillium Sans Pro",
                    "fontStyle": "normal",
                    "fontWeight": "600",
                  },
                ]
              }
              weight="Semibold"
            >
              Attributo non riconosciuto
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "#D2D6E3",
          "height": 0.5,
        }
      }
    />
  </View>
</View>
`;

exports[`ItwCredentialClaimsSection should match the snapshot when claims are visible 1`] = `
<View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      allowFontScaling={false}
      color="grey-700"
      defaultColor="black"
      defaultWeight="Semibold"
      font="TitilliumSansPro"
      fontStyle={
        {
          "fontSize": 18,
          "lineHeight": 25,
        }
      }
      style={
        [
          {
            "fontSize": 18,
            "lineHeight": 25,
          },
          {
            "color": "#555C70",
            "fontFamily": "Titillium Sans Pro",
            "fontStyle": "normal",
            "fontWeight": "600",
          },
        ]
      }
      weight="Semibold"
    >
      Personal data
    </Text>
    <View
      accessibilityLabel="Hide claim values"
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={8}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="toggle-claim-visibility"
    >
      <View
        style={
          [
            {
              "height": 24,
              "width": 24,
            },
            {
              "alignItems": "center",
            },
            {
              "justifyContent": "center",
            },
            {
              "transform": [
                {
                  "scale": undefined,
                },
              ],
            },
          ]
        }
      >
        <RNSVGSvgView
          accessibilityElementsHidden={true}
          accessibilityLabel=""
          accessible={false}
          align="xMidYMid"
          animatedProps={
            {
              "color": undefined,
            }
          }
          bbHeight={24}
          bbWidth={24}
          focusable={false}
          height={24}
          importantForAccessibility="no-hide-descendants"
          meetOrSlice={0}
          minX={0}
          minY={0}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "color": "#0073E6",
              },
              {
                "flex": 0,
                "height": 24,
                "width": 24,
              },
            ]
          }
          tintColor="#0073E6"
          vbHeight={24}
          vbWidth={24}
          width={24}
        >
          <RNSVGGroup
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
          >
            <RNSVGPath
              clipRule={0}
              d="M2.377 11.667A10.997 10.997 0 0 1 12.001 6c4.138 0 7.744 2.285 9.623 5.666a.686.686 0 0 1 0 .667A10.997 10.997 0 0 1 12 18a10.997 10.997 0 0 1-9.623-5.666.686.686 0 0 1 0-.667zM12.001 4A12.997 12.997 0 0 0 .629 10.696a2.686 2.686 0 0 0 0 2.609A12.997 12.997 0 0 0 12 20c4.894 0 9.155-2.705 11.372-6.696.45-.812.45-1.797 0-2.609A12.997 12.997 0 0 0 12.001 4zM15 12a3 3 0 1 1-6 0 3 3 0 0 1 6 0zm2 0a5 5 0 1 1-10 0 5 5 0 0 1 10 0z"
              fill={
                {
                  "type": 2,
                }
              }
              fillRule={0}
              propList={
                [
                  "fill",
                  "fillRule",
                ]
              }
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
    </View>
  </View>
  <View>
    <View
      accessibilityLabel="Nome Luigi"
      accessible={true}
      style={
        {
          "marginHorizontal": -24,
          "paddingHorizontal": 24,
          "paddingVertical": 12,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <View
            accessibilityElementsHidden={false}
            accessible={true}
            importantForAccessibility="yes"
          >
            <Text
              allowFontScaling={false}
              color="grey-700"
              defaultColor="blue"
              defaultWeight="Bold"
              font="TitilliumSansPro"
              fontStyle={
                {
                  "fontSize": 14,
                  "lineHeight": 21,
                }
              }
              style={
                [
                  {
                    "fontSize": 14,
                    "lineHeight": 21,
                  },
                  {
                    "color": "#555C70",
                    "fontFamily": "Titillium Sans Pro",
                    "fontStyle": "normal",
                    "fontWeight": "400",
                  },
                ]
              }
              weight="Regular"
            >
              Nome
            </Text>
            <Text
              allowFontScaling={false}
              color="black"
              defaultColor="black"
              defaultWeight="Semibold"
              font="TitilliumSansPro"
              fontStyle={
                {
                  "fontSize": 18,
                  "lineHeight": 25,
                }
              }
              numberOfLines={2}
              style={
                [
                  {
                    "fontSize": 18,
                    "lineHeight": 25,
                  },
                  {
                    "color": "#0E0F13",
                    "fontFamily": "Titillium Sans Pro",
                    "fontStyle": "normal",
                    "fontWeight": "600",
                  },
                ]
              }
              weight="Semibold"
            >
              Luigi
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "#D2D6E3",
          "height": 0.5,
        }
      }
    />
    <View
      accessibilityLabel="Cognome Rossi"
      accessible={true}
      style={
        {
          "marginHorizontal": -24,
          "paddingHorizontal": 24,
          "paddingVertical": 12,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <View
            accessibilityElementsHidden={false}
            accessible={true}
            importantForAccessibility="yes"
          >
            <Text
              allowFontScaling={false}
              color="grey-700"
              defaultColor="blue"
              defaultWeight="Bold"
              font="TitilliumSansPro"
              fontStyle={
                {
                  "fontSize": 14,
                  "lineHeight": 21,
                }
              }
              style={
                [
                  {
                    "fontSize": 14,
                    "lineHeight": 21,
                  },
                  {
                    "color": "#555C70",
                    "fontFamily": "Titillium Sans Pro",
                    "fontStyle": "normal",
                    "fontWeight": "400",
                  },
                ]
              }
              weight="Regular"
            >
              Cognome
            </Text>
            <Text
              allowFontScaling={false}
              color="black"
              defaultColor="black"
              defaultWeight="Semibold"
              font="TitilliumSansPro"
              fontStyle={
                {
                  "fontSize": 18,
                  "lineHeight": 25,
                }
              }
              numberOfLines={2}
              style={
                [
                  {
                    "fontSize": 18,
                    "lineHeight": 25,
                  },
                  {
                    "color": "#0E0F13",
                    "fontFamily": "Titillium Sans Pro",
                    "fontStyle": "normal",
                    "fontWeight": "600",
                  },
                ]
              }
              weight="Semibold"
            >
              Rossi
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "#D2D6E3",
          "height": 0.5,
        }
      }
    />
    <View
      accessibilityLabel="EtÃ  Attributo non riconosciuto"
      accessible={true}
      style={
        {
          "marginHorizontal": -24,
          "paddingHorizontal": 24,
          "paddingVertical": 12,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <View
            accessibilityElementsHidden={false}
            accessible={true}
            importantForAccessibility="yes"
          >
            <Text
              allowFontScaling={false}
              color="grey-700"
              defaultColor="blue"
              defaultWeight="Bold"
              font="TitilliumSansPro"
              fontStyle={
                {
                  "fontSize": 14,
                  "lineHeight": 21,
                }
              }
              style={
                [
                  {
                    "fontSize": 14,
                    "lineHeight": 21,
                  },
                  {
                    "color": "#555C70",
                    "fontFamily": "Titillium Sans Pro",
                    "fontStyle": "normal",
                    "fontWeight": "400",
                  },
                ]
              }
              weight="Regular"
            >
              EtÃ 
            </Text>
            <Text
              allowFontScaling={false}
              color="black"
              defaultColor="black"
              defaultWeight="Semibold"
              font="TitilliumSansPro"
              fontStyle={
                {
                  "fontSize": 18,
                  "lineHeight": 25,
                }
              }
              numberOfLines={2}
              style={
                [
                  {
                    "fontSize": 18,
                    "lineHeight": 25,
                  },
                  {
                    "color": "#0E0F13",
                    "fontFamily": "Titillium Sans Pro",
                    "fontStyle": "normal",
                    "fontWeight": "600",
                  },
                ]
              }
              weight="Semibold"
            >
              Attributo non riconosciuto
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "#D2D6E3",
          "height": 0.5,
        }
      }
    />
  </View>
</View>
`;
